c7n/resources/sagemaker.py

```diff
@@ -57,6 +57,96 @@ def _augment(r):
             return list(filter(None, w.map(_augment, resources)))
 
 
+@resources.register('sagemaker-job')
+class SagemakerJob(QueryResourceManager):
+
+    class resource_type(object):
+        service = 'sagemaker'
+        enum_spec = ('list_training_jobs', 'TrainingJobSummaries', None)
+        detail_spec = (
+            'describe_training_job', 'TrainingJobName',
+            'TrainingJobName', None)
+        id = 'TrainingJobArn'
+        name = 'TrainingJobName'
+        date = 'CreationTime'
+        dimension = None
+        filter_name = None
+
+    filter_registry = filters
+    permissions = ('sagemaker:ListTags',)
+
+    def augment(self, resources):
+        def _augment(r):
+            client = local_session(self.session_factory).client('sagemaker')
+            tags = client.list_tags(
+                ResourceArn=r['TrainingJobArn'])['Tags']
+            r.setdefault('Tags', []).extend(tags)
+            return r
+
+        with self.executor_factory(max_workers=1) as w:
+            return list(filter(None, w.map(_augment, resources)))
+
+
+@resources.register('sagemaker-endpoint')
+class SagemakerEndpoint(QueryResourceManager):
+
+    class resource_type(object):
+        service = 'sagemaker'
+        enum_spec = ('list_endpoints', 'Endpoints', None)
+        detail_spec = (
+            'describe_endpoint', 'EndpointName',
+            'EndpointName', None)
+        id = 'EndpointArn'
+        name = 'EndpointName'
+        date = 'CreationTime'
+        dimension = None
+        filter_name = None
+
+    filter_registry = filters
+    permissions = ('sagemaker:ListTags',)
+
+    def augment(self, endpoints):
+        def _augment(e):
+            client = local_session(self.session_factory).client('sagemaker')
+            tags = client.list_tags(
+                ResourceArn=e['EndpointArn'])['Tags']
+            e.setdefault('Tags', []).extend(tags)
+            return e
+
+        with self.executor_factory(max_workers=1) as w:
+            return list(filter(None, w.map(_augment, endpoints)))
+
+
+@resources.register('sagemaker-endpoint-config')
+class SagemakerEndpointConfig(QueryResourceManager):
+
+    class resource_type(object):
+        service = 'sagemaker'
+        enum_spec = ('list_endpoint_configs', 'EndpointConfigs', None)
+        detail_spec = (
+            'describe_endpoint_config', 'EndpointConfigName',
+            'EndpointConfigName', None)
+        id = 'EndpointConfigArn'
+        name = 'EndpointConfigName'
+        date = 'CreationTime'
+        dimension = None
+        filter_name = None
+
+    filter_registry = filters
+    permissions = ('sagemaker:ListTags',)
+
+    def augment(self, endpoints):
+        def _augment(e):
+            client = local_session(self.session_factory).client('sagemaker')
+            tags = client.list_tags(
+                ResourceArn=e['EndpointConfigArn'])['Tags']
+            e.setdefault('Tags', []).extend(tags)
```

Comment:
setdefault in augment seems odd, there's zero chance it exists, else it would be in the augment vs e['Tags'] = tags .. or even better scratch the temp variable 'tags'

Reply:
this pattern seems common through all the augments.